beamer/agent/chain.py

```diff
@@ -572,6 +575,61 @@ def _timestamp_is_l1_finalized(
     return int(time.time()) > timestamp + target_chain_finality
 
 
+def maybe_prove(claim: Claim, context: Context) -> bool:
+    request = context.requests.get(claim.request_id)
+
+    assert request is not None, "Active claim for non-existent request"
+
+    # mainnet: 10, goerli: 420, local: 17
+    if context.target_chain_id not in [10, 420, 17]:
+        return False
+
+    own_challenge_stake = Wei(claim.get_challenger_stake(context.address))
+    can_prove = (
+        claim.claimer == context.address and claim.challenger_exists()
+    ) or own_challenge_stake > 0
+
+    if not can_prove:
+        return False
+
+    if claim.message_proved:
+        return False
```

Comment:
This check should be put together with the one regarding the chain ID and moved to top of this function, see my comment about `_l1_resolution_threshold_reached`.

Reply:
Hmm... thinking more about this... perhaps it would be best if we made it so that this `maybe_prove` makes all messages "trivially" proved for all chains except Optimism?

E.g. pseudocode
```
def maybe_prove(...):
   if not optimism:
      claim.message_proved = True
      return

  ... all this Optimism related stuff...
```
and then in `process_claims` we could just
```python
           maybe_prove(claim, context)
           if _l1_resolution_threshold_reached(claim, context):
                maybe_resolve(claim, context)
```
To me that would look a little better because all chains except Optimism are proved by making a fill, it's just that for Optimism we have an extra step. But we stuff that inside `maybe_prove` and we don't need to worry about it.
(Future chains might also have different requirements.)

Then `maybe_resolve` should actually always check whether the message is proved.